File: sydney/Kernel/Client2/Connection.cpp
// -*-Mode: C++; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// Connection.cpp -- クライアントコネクション関連の関数定義
// 
// Copyright (c) 2006, 2013, 2015, 2023 Ricoh Company, Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// 

namespace {
const char srcFile[] = __FILE__;
const char moduleName[] = "Client2";
}

#include "SyDefault.h"
#include "SyReinterpretCast.h"
#include "Client2/Connection.h"
#include "Client2/DataSource.h"
#include "Client2/Port.h"
#ifdef SYD_COVERAGE
#include "Client2/Parameter.h"
#endif

#include "Common/Request.h"
#include "Common/IntegerData.h"
#include "Common/UnsignedIntegerData.h"
#include "Common/StringData.h"
#include "Common/IntegerArrayData.h"

#include "Communication/ConnectionSlaveID.h"

#include "Os/AutoCriticalSection.h"

#include "ModAutoPointer.h"
#ifdef SYD_COVERAGE
#include "ModOsDriver.h"
#endif

_TRMEISTER_USING
_TRMEISTER_CLIENT2_USING

namespace {

#ifdef SYD_COVERAGE

//	VARIABLE local
//	_$$::_iOpenWaitTime -- ポートのオープンを待つ時間 ( DEBUG 用)

ParameterInteger _iOpenWaitTime("Client2_OpenWaitTime", 0);

#endif

}

//	FUNCTION public
//	Client2::Connection::Connection -- コンストラクタ
//
//	NOTES
//
//	ARGUMENTS
//	Client2::DataSource&	cDataSource_
//		属するデータソースオブジェクト
//	Client2::Port*			pPort_
//		サーバ側のコネクションスレッドとの通信ポート
//
//	RETURN
//	なし
//
//	EXCEPTIONS

Connection::Connection(DataSource&	cDataSource_,
					   Port*		pPort_)
	: Object(Object::Type::Connection),
	  m_cDataSource(cDataSource_),
	  m_pPort(pPort_)
{
}

//	FUNCTION public
//	Client2::Connection::~Connection -- デストラクタ
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS

Connection::~Connection()
{
}

//	FUNCTION public
//	Client2::Connection::close -- クローズする
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS

void
Connection::close()
{
	if (m_pPort != 0) {

		try {

			// [<-] リクエスト
			const Common::Request	cRequest(Common::Request::EndConnection);
			m_pPort->writeObject(&cRequest);
			m_pPort->flush();
			// [->] ステータス
			m_pPort->readStatus();

			//コネクションをクローズする
			m_pPort->close();

		} catch (...) {
			Common::Thread::resetErrorCondition();
			// 例外は無視する
		}

		m_pPort->release(), m_pPort = 0;
	}
}

//	FUNCTION public
//	Client2::Connection::beginWorker -- ワーカを起動する
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	Client2::Port*
//		ワーカとの通信ポート
//
//	EXCEPTIONS

Port*
Connection::beginWorker()
{
	// コネクションプールからコネクションを得る
	ModAutoPointer<Port>	pPort = m_cDataSource.popPort();

	int	iSlaveID = Communication::ConnectionSlaveID::Any;
	ModAutoPointer<Common::IntegerData>	pSlaveID;
	ModAutoPointer<Common::IntegerData>	pWorkerID;

	try {

		if (pPort.get() != 0) iSlaveID = pPort->getSlaveID();


		{
			Os::AutoCriticalSection	cAuto(m_cPortLatch);

			// [<-] リクエスト
			const Common::Request	cRequest(Common::Request::BeginWorker);
			m_pPort->writeObject(&cRequest);
			// [<-] スレーブ ID
			const Common::IntegerData	cSlaveID(iSlaveID);
			m_pPort->writeObject(&cSlaveID);
			m_pPort->flush();

			// [->] スレーブ ID
			pSlaveID = m_pPort->readIntegerData();
			// [->] ワーカ ID
			pWorkerID = m_pPort->readIntegerData();
			// [->] ステータス
			m_pPort->readStatus();
		}

	} catch (...) {

		if (pPort.get() != 0 ) {
			if (pPort->isReuse()) {
				m_cDataSource.pushPort(pPort.release());
			} else {
				pPort->close();
			}
		}
		throw;
	}

	if (iSlaveID == Communication::ConnectionSlaveID::Any) {

		// 新しい通信ポート
		pPort = m_cDataSource.getNewPort(pSlaveID->getValue());
#ifdef SYD_COVERAGE
		ModOsDriver::Thread::sleep(_iOpenWaitTime.get());
#endif
		pPort->open();
		
		// 共通鍵を渡す(暗号化対応)
		pPort->setKey(m_cDataSource.getKey());
	} else {

		// プールされている通信ポート -> 同期を取る
		pPort->sync();
	}

	// ワーカ ID を設定
	pPort->setWorkerID(pWorkerID->getValue());

	return pPort.release();
}

//	FUNCTION public
//	Client2::Connection::cancelWorker -- ワーカをキャンセルする
//
//	NOTES
//
//	ARGUMENTS
//	int	iWorkerID_
//		キャンセルするワーカの ID
//
//	RETURN
//	なし
//
//	EXCEPTIONS

void
Connection::cancelWorker(int	iWorkerID_)
{
	Os::AutoCriticalSection	cAuto(m_cPortLatch);

	// [<-] リクエスト
	const Common::Request	cRequest(Common::Request::CancelWorker);
	m_pPort->writeObject(&cRequest);
	// [<-] ワーカ ID
	const Common::IntegerData	cWorkerID(iWorkerID_);
	m_pPort->writeObject(&cWorkerID);
	m_pPort->flush();

	// [->] ステータス
	m_pPort->readStatus();
}

//	FUNCTION public
//	Client2::Connection::erasePrepareStatement -- プリペアステートメントを削除する
//
//	NOTES
//
//	ARGUMENTS
//	const ModUnicodeString&	cstrDatabaseName_
//		データベース名
//	int						iPrepareID_
//		プリペアステートメント ID
//
//	RETURN
//	なし
//
//	EXCEPTIONS

void
Connection::erasePrepareStatement(	const ModUnicodeString&	cstrDatabaseName_,
									int						iPrepareID_)
{
	Os::AutoCriticalSection	cAuto(m_cPortLatch);

	// [<-] リクエスト
	const Common::Request	cRequest(Common::Request::ErasePrepareStatement);
	m_pPort->writeObject(&cRequest);
	// [<-] データベース名
	const Common::StringData	cDatabaseName(cstrDatabaseName_);
	m_pPort->writeObject(&cDatabaseName);
	// [<-] プリペアステートメント ID
	const Common::IntegerData	cPrepareID(iPrepareID_);
	m_pPort->writeObject(&cPrepareID);
	m_pPort->flush();

	// [->] ステータス
	m_pPort->readStatus();
}

//	FUNCTION public
//	Client2::Connection::disconnectPort -- 使用しない通信ポートを切断する
//
//	NOTES
//
//	ARGUMENTS
//	const ModVector<int>&	veciSlaveID_
//		切断する通信ポートのスレーブ ID の配列
//
//	RETURN
//	なし
//
//	EXCEPTIONS

void
Connection::disconnectPort(const ModVector<int>&	veciSlaveID_)
{
	Os::AutoCriticalSection	cAuto(m_cPortLatch);

	// [<-] リクエスト
	const Common::Request	cRequest(Common::Request::NoReuseConnection);
	m_pPort->writeObject(&cRequest);
	// [<-] スレーブ ID の配列
	const Common::IntegerArrayData	cSlaveID(veciSlaveID_);
	m_pPort->writeObject(&cSlaveID);
	m_pPort->flush();

	// [->] ステータス
	m_pPort->readStatus();
}

//	FUNCTION public
//	Client2::Connection::beginConnection -- 新しいコネクションを得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	Client2::Connection*
//		新しいクライアントコネクション
//
//	EXCEPTIONS

Connection*
Connection::beginConnection()
{
	Os::AutoCriticalSection	cAuto(m_cPortLatch);

	// [<-] リクエスト
	const Common::Request	cRequest(Common::Request::BeginConnection);
	m_pPort->writeObject(&cRequest);
	m_pPort->flush();

	// [->] スレーブ ID
	ModAutoPointer<Common::IntegerData>	pSlaveID = m_pPort->readIntegerData();

	// 新しい通信ポートを得る
	ModAutoPointer<Port>	pPort = m_cDataSource.getNewPort(pSlaveID->getValue());
	pPort->open();

	// [->] ステータス
	m_pPort->readStatus();

	// [->] 新しい通信ポートからステータスを得る
	pPort->readStatus();

	return new Connection(m_cDataSource, pPort.release());
}

//	FUNCTION public
//	Client2::Connection::isServerAvailable -- サーバ利用可能性を得る
//
//	NOTES
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	bool
//		サーバが利用可能な場合は true 、利用不可能な場合は false
//
//	EXCEPTIONS

bool
Connection::isServerAvailable()
{
	Os::AutoCriticalSection	cAuto(m_cPortLatch);

	// [<-] リクエスト
	Common::Request	cRequest(Common::Request::CheckAvailability);
	m_pPort->writeObject(&cRequest);
	// [<-] チェック対象
	Common::IntegerData	cTarget(Common::Request::AvailabilityTarget::Server);
	m_pPort->writeObject(&cTarget);
	m_pPort->flush();
	
	// [->] チェック結果
	ModAutoPointer<Common::IntegerData>	pResult = m_pPort->readIntegerData();
	// [->] ステータス
	m_pPort->readStatus();

	return (pResult->getValue() == 1) ? true : false;
}

//	FUNCTION public
//	Client2::Connection::isDatabaseAvailable -- データベース利用可能性を得る
//
//	NOTES
//
//	ARGUMENTS
//	Client2::Database::ID	iID_
//		データベース ID
//
//	RETURN
//	bool
//		データベースが利用可能な場合は true 、利用不可能な場合は false
//
//	EXCEPTIONS

bool
Connection::isDatabaseAvailable(Database::ID	iID_)
{
	Os::AutoCriticalSection	cAuto(m_cPortLatch);

	// [<-] リクエスト
	Common::Request	cRequest(Common::Request::CheckAvailability);
	m_pPort->writeObject(&cRequest);
	// [<-] チェック対象
	Common::IntegerData	cTarget(Common::Request::AvailabilityTarget::Database);
	m_pPort->writeObject(&cTarget);
	// [<-] データベース ID
	Common::UnsignedIntegerData cID(iID_);
	m_pPort->writeObject(&cID);
	m_pPort->flush();
	
	// [->] チェック結果
	ModAutoPointer<Common::IntegerData>	pResult = m_pPort->readIntegerData();
	// [->] ステータス
	m_pPort->readStatus();

	return (pResult->getValue() == 1) ? true : false;
}

//
//	Copyright (c) 2006, 2013, 2015, 2023 Ricoh Company, Ltd.
//	All rights reserved.
//

File: sydney/Python/src/doquedb/client/connection.py
# Copyright (c) 2024 Ricoh Company, Ltd.
#
# Licensed under the Apache License, Version 2.0 (the License);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
connection.py -- クライアントコネクション
"""
# TODO:is_database_availableの実装
from typing import Optional, List, TYPE_CHECKING
if TYPE_CHECKING:
    from .datasource import DataSource

from .object import Object
from .port import Port
from ..exception import exceptions
from ..common.serialdata import Request
from ..port.constants import ConnectionSlaveID
from ..common.arraydata import IntegerArrayData
from ..common.scalardata import IntegerData


class Connection(Object):
    """クライアントコネクションクラス.

    サーバとのコネクションを管理する. サーバ側にはこのクライアントコネクション１つごとにスレッドが存在している.
    """

    def __init__(self, datasource: 'DataSource', port: Port) -> None:
        super().__init__(Object.CONNECTION)
        self.__datasource = datasource
        self.__port: Optional[Port] = port

    def begin_connection(self) -> 'Connection':
        """新しいコネクションを得る.

        Returns:
            Connection: 新しいコネクション
        """
        assert self.__port is not None

        # [<-] リクエスト
        self.__port.write_object(
            Request(Request.request_map['BEGIN_CONNECTION']))
        self.__port.flush()
        # [->] スレーブIDを受け取る
        slave_data = self.__port.read_integerdata()

        # 新しい通信ポートを得る
        port = self.__datasource.new_port(slave_data.value)
        port.open()

        # [->] ステータス
        self.__port.read_status()
        # [->] 新しい通信ポートからステータスを得る
        port.read_status()

        return Connection(self.__datasource, port)

    def begin_worker(self) -> Port:
        """ワーカを起動する.

        Returns:
            Port: 接続ポート

        Raises:
            UnexpectedError: データソースから持ってきたポートが開放されていなかった場合

        Notes:
            ポートプールにポートがない場合、slaveID=ANYとしてDoqueDBから返ってくるポートを使用する.

        """
        assert self.__port is not None
        # ポートプールからポートを得る
        port = self.__datasource.pop_port()
        slave_id = ConnectionSlaveID.ANY.value

        try:
            # datasoureにportが登録されていればそれを用いる
            if port:
                slave_id = port.slave_id
                # slave_idがANYだった場合例外をあげる
                if slave_id == ConnectionSlaveID.ANY.value:
                    raise exceptions.UnexpectedError(
                        'tried to use invalid port from datasource')

            # [<-] リクエスト
            self.__port.write_object(
                Request(Request.request_map['BEGIN_WORKER']))
            # [<-] スレーブID
            self.__port.write_object(IntegerData(slave_id))
            self.__port.flush()

            # [->] スレーブID
            slave_data = self.__port.read_integerdata()
            # [->] ワーカID
            worker_data = self.__port.read_integerdata()
            # [->] ステータス
            self.__port.read_status()
        except Exception as err:
            if port:
                if err is exceptions.UnexpectedError:
                    # ConnectionSlaveIDがANYだった場合portは返さない
                    port.close()
                else:
                    # popしたportを戻す
                    self.__datasource.push_port(port)
            raise err

        if slave_id == ConnectionSlaveID.ANY.value:
            # 新しい通信ポート
            port = self.__datasource.new_port(slave_data.value)
            port.open()
        assert port is not None
        # ワーカIDを設定
        port.worker_id = worker_data.value

        return port

    def cancel_worker(self, worker_id: int) -> None:
        """ワーカをキャンセルする.

        Args:
            worker_id (int): キャンセルするワーカのID
        """
        assert self.__port is not None
        # [<-] リクエスト
        self.__port.write_object(Request(Request.request_map['CANCEL_WORKER']))
        # [<-] ワーカID
        self.__port.write_object(IntegerData(worker_id))
        self.__port.flush()

        # [->] ステータス
        self.__port.read_status()

    def disconnect_port(self, slave_id: List[IntegerData]) -> None:
        """使用しない通信ポートを切断する.

        Args:
            slave_id (List[int]): 切断するポートのスレーブIDのリスト
        """
        assert self.__port is not None
        # [<-] リクエスト
        self.__port.write_object(
            Request(Request.request_map['NO_REUSE_CONNECTION']))
        # [<-] スレーブIDの配列
        self.__port.write_object(IntegerArrayData(slave_id))
        self.__port.flush()

        # [->] ステータス
        self.__port.read_status()

    def is_serever_available(self) -> bool:
        """サーバの利用可能性を得る.

        Returns:
            bool: サーバが利用可能な場合はTrue, 利用不可能な場合はFalse
        """
        assert self.__port is not None
        # [<-] リクエスト
        self.__port.write_object(
            Request(Request.request_map['CHECK_AVAILABILITY']))
        # [<-] チェック対象
        self.__port.write_object(IntegerData(
            Request.request_map['AVAILABILITY_TARGET_SERVER']))

        # [->] チェック結果
        result = self.__port.read_integerdata()
        # [->] ステータス
        self.__port.read_status()

        return result.value == 1

    def is_database_available(self) -> bool:
        """任意のデータベースの利用可能性を得る.

        Returns:
            bool: データベースが利用可能な場合はTrue , 利用不可能な場合はFalse
        """
        # TODO: 実装
        pass

    def close(self) -> None:
        """クローズする.

        Notes:
            このメソッドは例外を送出しない.
        """
        if self.__port is None:
            return

        try:
            # [<-] リクエスト
            self.__port.write_object(
                Request(Request.request_map['END_CONNECTION']))
            self.__port.flush()
            # [->] ステータス
            self.__port.read_status()
        except Exception:
            # 例外は無視する
            pass

        try:
            self.__port.close()
        except Exception:
            # 例外は無視する
            pass
        self.__port = None
